discoverd/client: Cluster awareness/failover

josephglanville opened the pull request:
This changeset makes the discoverd client cluster aware.

It can now be initialised with a connection string of comma delimited servers.
The first server is considered to be preferred and it will use it unless it's unavailable.
If it does become unavailable then the rest of the servers configured will be tried until a working one is found, this server will then be pinned for 60 seconds, after the TTL expires the client will attempt to use the preferred server once more.
Writes are directed to the leader by recording where redirects are sent, if at any point the leader becomes unavailable we will try issue the write to another server and have it redirect us to the new leader.

Closes #2091

lmars commented on discoverd/client/client.go:
```diff
@@ -1,261 +1,192 @@
-	if url == "" {
-		url = "http://127.0.0.1:1111"
+func NewClientWithConfig(config Config) *Client {
+	client := Client{}
```
We tend to write `var client Client`

lmars replied:
Then again, looking at the next lines, this should probably be:

```go
client := &Client{
	servers: make([]*httpclient.Client, 0, len(config.Endpoints)),
}
```

josephglanville replied:
Yep this makes more sense.

lmars commented on discoverd/client/client.go:
```diff
@@ -1,261 +1,192 @@
+		if !strings.HasPrefix(u, "http") {
+			u = "http://" + u
+		}
+		formatted = append(formatted, u)
```
Given you have created a slice with a non-zero length, this either needs to index into that slice (i.e. `formatted[i] = u`), or just create the slice with a capacity but zero length.

josephglanville replied:
I see.. will fix.

lmars commented on discoverd/client/client.go:
```diff
@@ -1,261 +1,192 @@
-
-type ServiceConfig struct {
-	LeaderType LeaderType `json:"leader_type"`
+func (c *Client) updateLeader(url string) {
```
This argument seems to be a `host`, rather than a `url`

josephglanville replied:
Yep, rename of arg in order.

lmars commented on discoverd/client/client.go:
```diff
@@ -1,261 +1,192 @@
+		hc := c.servers[k]
+		err := hc.Send(method, path, in, out)
+		if err != nil {
+			continue
```
We should save these errors so they can be returned if all the hosts fail

josephglanville replied:
Hmm, would the best approach here be to collect these into a slice and then return them as a big Error? I could probably create a bunch of custom errors like ErrNoHosts etc to handle the different ways we can blow up.

lmars replied:
I think I'd like to see an error like:

```
error sending HTTP request: 10.0.0.1:1111 connection refused, 10.0.0.2:1111 something went wrong, 10.0.0.3:1111 connection reset
```

josephglanville replied:
Yep, that is what I was thinking.

josephglanville commented:
So right now it will only failover to another server if the error is a `net.OpError`.
I think it should also failover if there is no current Raft leader because there could be a net split in play and if we keep trying peers we might find a peer on the majority side of the net split.

lmars commented on discoverd/client/client.go:
```diff
@@ -1,266 +1,205 @@
-			},
-		},
-	}
+	return NewClientWithConfig(Config{Endpoints: formatURLs([]string{url})})
```
`formatURLs` should take `urls ...string` so you can call it here like `formatURLs(url)`

lmars commented on discoverd/client/client.go:
```diff
@@ -1,266 +1,205 @@
+		return true
 	}
+	return false
 }
```
```go
_, ok := err.(*net.OpError)
return ok
```

lmars commented on host/discoverd.go:
```diff
@@ -78,6 +78,7 @@ func (d *DiscoverdManager) ConnectLocal(url string) error {
 	}
 	d.local.Store(true)
 
+	fmt.Println("Setting default env var DISCOVERD=", url)
```
Either drop this or use `log15`

josephglanville commented:
This is ready for a review pass.

Failing over on `ErrNoKnownLeader` will probably happen in a future PR as I don't want to touch the server much in this pass. (supporting it properly requires converting the server to use JSONError most likely)

titanous commented on discoverd/client/client.go:
```diff
@@ -1,303 +1,251 @@
+			return nil, err
+		}
+		if time.Since(pinUpdated) > PinTTL {
+			k = 0 // TTL has timed out, use prefered server on next request
```
preferred

titanous commented:
This appears to rely on all available discoverd servers to be specified ahead of time. Will another pass implement getting an updated list from the server?

josephglanville commented:
Yes, I think when I work out what strategy will be used to update the list of peers for when we do reconfigurations etc I will also use it to have the client automatically keep the list of servers up to date.
In all likelihood it will probably be finding a good way to make something like `/raft/peers` stream changes. Potentially we could poll but that seems nasty given how good our tools for long-poll/streaming are.

josephglanville commented:
@titanous this is ready for some review. I implemented updates to the peer set by setting a HTTP header `DISCOVERD_PEERS` on each response that the client can use to keep in sync. This avoids the need for a goroutine.

titanous commented on discoverd/client/client.go:
```diff
@@ -36,268 +26,256 @@ func init() {
+		} else if err != nil {
+			return nil, err
+		}
+		// If the pinned server failed to fufil our request then update the pin
```
s/fufil/fulfill/

titanous commented on discoverd/client/client.go:
```diff
@@ -36,268 +26,256 @@ func init() {
+			c.updatePin(k)
+		}
+		if peers := rsp.Header.Get("DISCOVERD_PEERS"); peers != "" {
+			c.updateServers(strings.Split(peers, ","))
```
Should this include the raft index so that out of order request response processing doesn't nuke a new peer list?

josephglanville commented:
@titanous comments addressed. Added the index to prevent changes to peer requests from older indices. Also modified how the servers are maintained and updated to prevent creating new httpclient.Clients if the transport config is the same.

titanous commented on discoverd/client/client.go:
```diff
@@ -36,268 +27,306 @@ func init() {
+	return &httpclient.Client{
+		URL: url,
+		HTTP: &http.Client{
+			CheckRedirect: checkRedirect,
```
Can this `http.Client` client be defined once globally? It doesn't look like it varies based on the URL.

josephglanville replied:
I don't see why not, I wasn't aware you could share them across httpclient.Client instances but if so  may as well.

titanous replied:
Yeah, I don't think there's any reason we can't do that.

titanous commented on discoverd/server/handler.go:
```diff
@@ -108,15 +110,25 @@ func proxyWhitelisted(r *http.Request) bool {
+		if err == nil {
+			w.Header().Set("Discoverd-Current-Peers", strings.Join(peers, ","))
+		}
+		w.Header().Set("Discoverd-Current-Index", strconv.FormatUint(h.Store.LastIndex(), 10))
```
Should we drop `-Current` from these so that it's just `Discoverd-Peers` and `Discoverd-Index`?

josephglanville replied:
Yeah I guess current is redundant. It wouldn't be sending them to you if it didn't think they were current.

titanous commented:
LGTM

temujin9 commented:
Thanks, guys. LMK when this drops, so I can start testing it out on our sandbox.

titanous commented:
@temujin9 This is in `v20160602.0`.
